### PR TITLE
Use ZMQ for passing microbatches between processes

### PR DESCRIPTION
**Pull request by usmanm**

### CODE REVIEW

File: src/Makefile.global.in [hunk 1]
```diff
@@ -598,6 +598,13 @@ TAS         = @TAS@
 
 ##########################################################################
 #
+# Link to ZMQ
+
+LIBS := -lzmq $(LIBS)
```
Comment (derekjn) [inline src/Makefile.global.in]: We should statically link this.

File: src/backend/nodes/print.c [hunk 2]
```diff
@@ -30,7 +30,7 @@
 #include "parser/parsetree.h"
 #include "utils/lsyscache.h"
 
-
+#define SLEEP_ON_ASSERT
```
Comment (derekjn) [inline src/backend/nodes/print.c]: Unintentional?

File: src/backend/utils/misc/guc.c [hunk 3]
```diff
@@ -2784,16 +2763,27 @@ static struct config_int ConfigureNamesInt[] =
 	},
 
 	{
-		{"continuous_query_batch_size", PGC_SIGHUP, QUERY_TUNING,
+		{"continuous_query_num_batch", PGC_SIGHUP, QUERY_TUNING,
```
Comment (derekjn) [inline src/backend/utils/misc/guc.c]: This name seems slightly unintuitive. Maybe we should call these,

    continuous_query_batch_length
    continuous_query_batch_size

File: src/test/regress/expected/pipeline_stats.out [hunk 4]
```diff
@@ -54,7 +54,7 @@ SELECT output_rows, errors, cv_create, cv_drop FROM pipeline_stats WHERE type =
 SELECT input_rows, input_bytes, errors FROM pipeline_stats WHERE type = 'worker';
  input_rows | input_bytes | errors 
 ------------+-------------+--------
-       2000 |      316000 |      0
+       2000 |       96000 |      0
```
Comment (derekjn) [inline src/test/regress/expected/pipeline_stats.out]: :clap:

File: src/Makefile.global.in (hunk 1 above)
Comment (usmanm) [inline src/Makefile.global.in]: ZeroMQ is in C++ so I'm not sure how we can statically link it without compiling with `g++`.

Comment (derekjn) [inline src/Makefile.global.in]: I guess for a development build using it as a regular library is fine. But for packages I don't see any reason not to build it as a static library with `g++`.

Comment (derekjn) [inline src/Makefile.global.in]: I've created a ticket for this and will do it since I'm creating the new packaging containers right now.

Comment (usmanm) [inline src/Makefile.global.in]: https://gist.github.com/usmanm/32a54a6b0f1f29d7737f86e29f837afa

This builds a static library as well.

**Comment by usmanm** [comment]
```
TRAP: BadArgument("!(((context) != ((void *)0) && (((((const Node*)((context)))->type) == T_AllocSetContext))))", File: "mcxt.c", Line: 679, PID: 8078, Query: concurrent_copy0)
Assertion failure (PID 8078)
version: PipelineDB  at revision  on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 5.4.0-6ubuntu1~16.04.1) 5.4.0 20160609, 64-bit
backtrace:
pipelinedb: bgworker: combiner1 [pipeline]   (ExceptionalCondition+0xc3)[0x9aa5ab]
pipelinedb: bgworker: combiner1 [pipeline]   (MemoryContextAlloc+0x40)[0x9da5fd]
pipelinedb: bgworker: combiner1 [pipeline]   (PushActiveSnapshot+0x40)[0x9f48d0]
pipelinedb: bgworker: combiner1 [pipeline]   (PortalStart+0x1c9)[0x849bab]
pipelinedb: bgworker: combiner1 [pipeline]   [0x79dfd2]
pipelinedb: bgworker: combiner1 [pipeline]   [0x7a0df2]
pipelinedb: bgworker: combiner1 [pipeline]   (ContinuousQueryCombinerMain+0x195)[0x7a1b56]
pipelinedb: bgworker: combiner1 [pipeline]   [0x79b6ac]
pipelinedb: bgworker: combiner1 [pipeline]   (StartBackgroundWorker+0x2ac)[0x7ad4d1]
pipelinedb: bgworker: combiner1 [pipeline]   [0x7c23c7]
pipelinedb: bgworker: combiner1 [pipeline]   [0x7c26b0]
pipelinedb: bgworker: combiner1 [pipeline]   [0x7c16e0]
/lib/x86_64-linux-gnu/libc.so.6(+0x354a0)[0x7fa4806f44a0]
/lib/x86_64-linux-gnu/libc.so.6(__select+0x13)[0x7fa4807bbcf3]
pipelinedb: bgworker: combiner1 [pipeline]   [0x7bcddf]
pipelinedb: bgworker: combiner1 [pipeline]   (PostmasterMain+0x1114)[0x7bc4d2]
pipelinedb: bgworker: combiner1 [pipeline]   [0x6e90cb]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)[0x7fa4806df830]
pipelinedb: bgworker: combiner1 [pipeline]   (_start+0x29)[0x469cb9]
LOG:  worker process: combiner1 [pipeline] (PID 8078) was terminated by signal 6: Aborted
```

**Comment by usmanm** [comment]
```
Segmentation fault (PID 10259)
version: PipelineDB  at revision  on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 5.4.0-6ubuntu1~16.04.1) 5.4.0 20160609, 64-bit
backtrace:
pipelinedb: bgworker: combiner1 [pipeline]   (debug_segfault+0x34)[0x725c95]
/lib/x86_64-linux-gnu/libc.so.6(+0x354a0)[0x7efcbbefa4a0]
pipelinedb: bgworker: combiner1 [pipeline]   [0x9d8d22]
pipelinedb: bgworker: combiner1 [pipeline]   (MemoryContextAlloc+0xd0)[0x9da6e1]
pipelinedb: bgworker: combiner1 [pipeline]   (MemoryContextCreate+0x7e)[0x9da48b]
pipelinedb: bgworker: combiner1 [pipeline]   (AllocSetContextCreate+0x41)[0x9d86ac]
pipelinedb: bgworker: combiner1 [pipeline]   (CreatePortal+0x165)[0x9dbca1]
pipelinedb: bgworker: combiner1 [pipeline]   [0x79df81]
pipelinedb: bgworker: combiner1 [pipeline]   [0x7a0e3c]
pipelinedb: bgworker: combiner1 [pipeline]   (ContinuousQueryCombinerMain+0x195)[0x7a1ba0]
pipelinedb: bgworker: combiner1 [pipeline]   [0x79b6f6]
pipelinedb: bgworker: combiner1 [pipeline]   (StartBackgroundWorker+0x2ac)[0x7ad51b]
pipelinedb: bgworker: combiner1 [pipeline]   [0x7c2411]
pipelinedb: bgworker: combiner1 [pipeline]   [0x7c26fa]
pipelinedb: bgworker: combiner1 [pipeline]   [0x7bd18f]
pipelinedb: bgworker: combiner1 [pipeline]   (PostmasterMain+0x1114)[0x7bc51c]
pipelinedb: bgworker: combiner1 [pipeline]   [0x6e910b]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)[0x7efcbbee5830]
```

Segfault around same place. Somehow the `MemoryContext` struct is getting corrupted...

**Comment by usmanm** [comment]
```
WARNING:  problem in alloc set QueryStateCxt: detected write past chunk end in block 0x16e8e70, chunk 0x16e8e90
LOG:  autovacuum launcher shutting down
LOG:  continuous query scheduler shutting down
FATAL:  terminating connection due to administrator command
LOG:  continuous query process "worker1 [pipeline]" was killed
LOG:  continuous query process "worker0 [pipeline]" was killed
*** Error in `pipelinedb: bgworker: combiner1 [pipeline]   ': free(): invalid next size (normal): 0x00000000016e8e70 ***
======= Backtrace: =========
/lib/x86_64-linux-gnu/libc.so.6(+0x77725)[0x7fc8e4483725]
/lib/x86_64-linux-gnu/libc.so.6(+0x7ff4a)[0x7fc8e448bf4a]
LOG:  worker process: worker1 [pipeline] (PID 11987) exited with exit code 1
LOG:  continuous query process "combiner0 [pipeline]" was killed
/lib/x86_64-linux-gnu/libc.so.6(cfree+0x4c)[0x7fc8e448fabc]
pipelinedb: bgworker: combiner1 [pipeline]   [0x9d8b7c]
pipelinedb: bgworker: combiner1 [pipeline]   (MemoryContextDelete+0xc3)[0x9d9dce]
pipelinedb: bgworker: combiner1 [pipeline]   (MemoryContextDeleteChildren+0x49)[0x9d9e26]
pipelinedb: bgworker: combiner1 [pipeline]   (MemoryContextDelete+0x91)[0x9d9d9c]
LOG:  worker process: worker0 [pipeline] (PID 11988) exited with exit code 1
pipelinedb: bgworker: combiner1 [pipeline]   (ContExecutorDestroy+0x20)[0x7a58aa]
pipelinedb: bgworker: combiner1 [pipeline]   (ContinuousQueryCombinerMain+0x40c)[0x7a1e17]
pipelinedb: bgworker: combiner1 [pipeline]   [0x79b6f6]
pipelinedb: bgworker: combiner1 [pipeline]   (StartBackgroundWorker+0x2ac)[0x7ad51b]
pipelinedb: bgworker: combiner1 [pipeline]   [0x7c2411]
pipelinedb: bgworker: combiner1 [pipeline]   [0x7c26fa]
LOG:  worker process: combiner0 [pipeline] (PID 11986) exited with exit code 1
pipelinedb: bgworker: combiner1 [pipeline]   [0x7bd18f]
pipelinedb: bgworker: combiner1 [pipeline]   (PostmasterMain+0x1114)[0x7bc51c]
pipelinedb: bgworker: combiner1 [pipeline]   [0x6e910b]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)[0x7fc8e442c830]
pipelinedb: bgworker: combiner1 [pipeline]   (_start+0x29)[0x469cf9]
======= Memory map: ========
00400000-00c80000 r-xp 00000000 08:01 2628310                            /home/usmanm/pipelinedb/src/backend/pipelinedb
00e7f000-00e80000 r--p 0087f000 08:01 2628310                            /home/usmanm/pipelinedb/src/backend/pipelinedb
00e80000-00e8d000 rw-p 00880000 08:01 2628310                            /home/usmanm/pipelinedb/src/backend/pipelinedb
00e8d000-00ed6000 rw-p 00000000 00:00 0 
0151a000-0156b000 rw-p 00000000 00:00 0                                  [heap]
0156b000-01aa2000 rw-p 00000000 00:00 0                                  [heap]
7fc8cc000000-7fc8cc021000 rw-p 00000000 00:00 0 
7fc8cc021000-7fc8d0000000 ---p 00000000 00:00 0 
7fc8d4000000-7fc8d40d7000 rw-p 00000000 00:00 0 
7fc8d40d7000-7fc8d8000000 ---p 00000000 00:00 0 
7fc8d8c16000-7fc8d8c17000 ---p 00000000 00:00 0 
7fc8d8c17000-7fc8d9417000 rw-p 00000000 00:00 0 
7fc8d9417000-7fc8d9418000 ---p 00000000 00:00 0 
7fc8d9418000-7fc8d9c18000 rw-p 00000000 00:00 0 
7fc8d9c18000-7fc8e3086000 rw-s 00000000 00:05 109995565                  /dev/zero (deleted)
7fc8e3086000-7fc8e3091000 r-xp 00000000 08:01 8000054                    /lib/x86_64-linux-gnu/libnss_files-2.23.so
7fc8e3091000-7fc8e3290000 ---p 0000b000 08:01 8000054                    /lib/x86_64-linux-gnu/libnss_files-2.23.so
7fc8e3290000-7fc8e3291000 r--p 0000a000 08:01 8000054                    /lib/x86_64-linux-gnu/libnss_files-2.23.so
7fc8e3291000-7fc8e3292000 rw-p 0000b000 08:01 8000054                    /lib/x86_64-linux-gnu/libnss_files-2.23.so
7fc8e3292000-7fc8e3298000 rw-p 00000000 00:00 0 
7fc8e3298000-7fc8e3c57000 r--p 00000000 08:01 8526877                    /usr/lib/locale/locale-archive
7fc8e3c57000-7fc8e3c6d000 r-xp 00000000 08:01 7999993                    /lib/x86_64-linux-gnu/libgcc_s.so.1
7fc8e3c6d000-7fc8e3e6c000 ---p 00016000 08:01 7999993                    /lib/x86_64-linux-gnu/libgcc_s.so.1
7fc8e3e6c000-7fc8e3e6d000 rw-p 00015000 08:01 7999993                    /lib/x86_64-linux-gnu/libgcc_s.so.1
7fc8e3e6d000-7fc8e3fdf000 r-xp 00000000 08:01 8521297                    /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21
7fc8e3fdf000-7fc8e41df000 ---p 00172000 08:01 8521297                    /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21
7fc8e41df000-7fc8e41e9000 r--p 00172000 08:01 8521297                    /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21
7fc8e41e9000-7fc8e41eb000 rw-p 0017c000 08:01 8521297                    /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21
7fc8e41eb000-7fc8e41ef000 rw-p 00000000 00:00 0 
7fc8e41ef000-7fc8e4207000 r-xp 00000000 08:01 8000101                    /lib/x86_64-linux-gnu/libpthread-2.23.so
7fc8e4207000-7fc8e4406000 ---p 00018000 08:01 8000101                    /lib/x86_64-linux-gnu/libpthread-2.23.so
7fc8e4406000-7fc8e4407000 r--p 00017000 08:01 8000101                    /lib/x86_64-linux-gnu/libpthread-2.23.so
7fc8e4407000-7fc8e4408000 rw-p 00018000 08:01 8000101                    /lib/x86_64-linux-gnu/libpthread-2.23.so
7fc8e4408000-7fc8e440c000 rw-p 00000000 00:00 0 
7fc8e440c000-7fc8e45cc000 r-xp 00000000 08:01 7999955                    /lib/x86_64-linux-gnu/libc-2.23.so
7fc8e45cc000-7fc8e47cb000 ---p 001c0000 08:01 7999955                    /lib/x86_64-linux-gnu/libc-2.23.so
7fc8e47cb000-7fc8e47cf000 r--p 001bf000 08:01 7999955                    /lib/x86_64-linux-gnu/libc-2.23.so
7fc8e47cf000-7fc8e47d1000 rw-p 001c3000 08:01 7999955                    /lib/x86_64-linux-gnu/libc-2.23.so
7fc8e47d1000-7fc8e47d5000 rw-p 00000000 00:00 0 
7fc8e47d5000-7fc8e48dd000 r-xp 00000000 08:01 8000025                    /lib/x86_64-linux-gnu/libm-2.23.so
7fc8e48dd000-7fc8e4adc000 ---p 00108000 08:01 8000025                    /lib/x86_64-linux-gnu/libm-2.23.so
7fc8e4adc000-7fc8e4add000 r--p 00107000 08:01 8000025                    /lib/x86_64-linux-gnu/libm-2.23.so
7fc8e4add000-7fc8e4ade000 rw-p 00108000 08:01 8000025                    /lib/x86_64-linux-gnu/libm-2.23.so
7fc8e4ade000-7fc8e4ae1000 r-xp 00000000 08:01 7999979                    /lib/x86_64-linux-gnu/libdl-2.23.so
7fc8e4ae1000-7fc8e4ce0000 ---p 00003000 08:01 7999979                    /lib/x86_64-linux-gnu/libdl-2.23.so
7fc8e4ce0000-7fc8e4ce1000 r--p 00002000 08:01 7999979                    /lib/x86_64-linux-gnu/libdl-2.23.so
7fc8e4ce1000-7fc8e4ce2000 rw-p 00003000 08:01 7999979                    /lib/x86_64-linux-gnu/libdl-2.23.so
7fc8e4ce2000-7fc8e4ce9000 r-xp 00000000 08:01 8000109                    /lib/x86_64-linux-gnu/librt-2.23.so
7fc8e4ce9000-7fc8e4ee8000 ---p 00007000 08:01 8000109                    /lib/x86_64-linux-gnu/librt-2.23.so
7fc8e4ee8000-7fc8e4ee9000 r--p 00006000 08:01 8000109                    /lib/x86_64-linux-gnu/librt-2.23.so
7fc8e4ee9000-7fc8e4eea000 rw-p 00007000 08:01 8000109                    /lib/x86_64-linux-gnu/librt-2.23.so
7fc8e4eea000-7fc8e4f4f000 r-xp 00000000 08:01 8551798                    /usr/lib/libzmq.so.5.0.1
7fc8e4f4f000-7fc8e514e000 ---p 00065000 08:01 8551798                    /usr/lib/libzmq.so.5.0.1
7fc8e514e000-7fc8e5152000 r--p 00064000 08:01 8551798                    /usr/lib/libzmq.so.5.0.1
7fc8e5152000-7fc8e5153000 rw-p 00068000 08:01 8551798                    /usr/lib/libzmq.so.5.0.1
7fc8e5153000-7fc8e5179000 r-xp 00000000 08:01 7999927                    /lib/x86_64-linux-gnu/ld-2.23.so
7fc8e5291000-7fc8e5353000 rw-p 00000000 00:00 0 
7fc8e5353000-7fc8e535a000 rw-p 00000000 00:00 0 
7fc8e5373000-7fc8e5374000 rw-p 00000000 00:00 0 
7fc8e5374000-7fc8e5375000 rw-s 00000000 00:05 318701588                  /SYSV6b8b4567 (deleted)
7fc8e5375000-7fc8e5376000 rw-s 00000000 00:05 318668815                  /SYSV02ef0d89 (deleted)
7fc8e5376000-7fc8e5378000 rw-p 00000000 00:00 0 
7fc8e5378000-7fc8e5379000 r--p 00025000 08:01 7999927                    /lib/x86_64-linux-gnu/ld-2.23.so
7fc8e5379000-7fc8e537a000 rw-p 00026000 08:01 7999927                    /lib/x86_64-linux-gnu/ld-2.23.so
7fc8e537a000-7fc8e537b000 rw-p 00000000 00:00 0 
7ffed0af5000-7ffed0b1a000 rw-p 00000000 00:00 0                          [stack]
7ffed0b49000-7ffed0b4b000 r--p 00000000 00:00 0                          [vvar]
7ffed0b4b000-7ffed0b4d000 r-xp 00000000 00:00 0                          [vdso]
ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0                  [vsyscall]
```

**Comment by usmanm** [comment]
Out of window firing twice? https://circleci.com/gh/pipelinedb/pipelinedb/5364
Proc stats 3 instead of 4? https://circleci.com/gh/pipelinedb/pipelinedb/5365
Aforementioned memory corruptions: https://circleci.com/gh/pipelinedb/pipelinedb/5365
Restart recovery not showing updates? https://circleci.com/gh/pipelinedb/pipelinedb/5367

**Comment by usmanm** [comment]
`WARNING:  registered snapshots seem to remain after cleanup`

See which process isn't doing proper clean up

**Pull request closed without merging**